hoodie-common/src/main/java/com/uber/hoodie/common/table/log/AbstractHoodieLogRecordScanner.java

```diff
@@ -0,0 +1,312 @@
+  // Log File Paths
+  private final List<String> logFilePaths;
+  // Read Lazily flag
+  private final boolean readBlocksLazily;
+  // Reverse reader - Not implemented yet (NA -> Why do we need ?)
+  // but present here for plumbing for future implementation
+  private final boolean reverseReader;
+  // FileSystem
+  private final FileSystem fs;
+  // Store the last instant log blocks (needed to implement rollback)
+  private Deque<HoodieLogBlock> currentInstantLogBlocks = new ArrayDeque<>();
+  // Progress
+  private float progress = 0.0f;
+
+  public AbstractHoodieLogRecordScanner(FileSystem fs, String basePath, List<String> logFilePaths,
+      Schema readerSchema, String latestInstantTime,
+      boolean readBlocksLazily, boolean reverseReader) {
+    this.readerSchema = readerSchema;
+    this.latestInstantTime = latestInstantTime;
+    this.hoodieTableMetaClient = new HoodieTableMetaClient(fs.getConf(), basePath);
+    // load class from the payload fully qualified class name
+    this.payloadClassFQN = this.hoodieTableMetaClient.getTableConfig().getPayloadClass();
+    this.logFilePaths = logFilePaths;
+    this.readBlocksLazily = readBlocksLazily;
+    this.reverseReader = reverseReader;
+    this.fs = fs;
+  }
+
+  /**
+   * Scan Log files
+   */
+  public void scan() {
+    try {
+      // iterate over the paths
+      HoodieLogFormatReader logFormatReaderWrapper =
+          new HoodieLogFormatReader(fs,
+              logFilePaths.stream().map(logFile -> new HoodieLogFile(new Path(logFile)))
+                  .collect(Collectors.toList()), readerSchema, readBlocksLazily, reverseReader);
+      Set<HoodieLogFile> seenLogFiles = new HashSet<>();
+      while (logFormatReaderWrapper.hasNext()) {
+        HoodieLogFile logFile = logFormatReaderWrapper.getLogFile();
+        log.info("Scanning log file " + logFile);
+        seenLogFiles.add(logFile);
+        totalLogFiles.set(seenLogFiles.size());
+        progress = seenLogFiles.size() - 1 / logFilePaths.size();
+        // Use the HoodieLogFileReader to iterate through the blocks in the log file
+        HoodieLogBlock r = logFormatReaderWrapper.next();
+        totalLogBlocks.incrementAndGet();
+        if (r.getBlockType() != CORRUPT_BLOCK
+            && !HoodieTimeline.compareTimestamps(r.getLogBlockHeader().get(INSTANT_TIME),
+            this.latestInstantTime,
+            HoodieTimeline.LESSER_OR_EQUAL)) {
+          //hit a block with instant time greater than should be processed, stop processing further
+          break;
+        }
+        switch (r.getBlockType()) {
+          case AVRO_DATA_BLOCK:
+            log.info("Reading a data block from file " + logFile.getPath());
+            if (isNewInstantBlock(r) && !readBlocksLazily) {
+              // If this is an avro data block belonging to a different commit/instant,
+              // then merge the last blocks and records into the main result
+              processNextCommitBlocks(currentInstantLogBlocks);
```

Comment:
This is actually processing the blocks seen for the last commit, the name might be misleading, can we rename this to `processSeenCommitBlocks` or something ?

Reply:
Thanks. Sure, will do